### DISCUSSION TITLE
Render heatmap in MVT Layer

### DISCUSSION THREAD

**Discussion by makeupsomething** [visgl/deck.gl#5098] [Q&A]
Using `react-map-gl` I can load a `<Source/>` and inside that I can have a `<Layer type="heatmap" />` to show a heatmap. 
Is this also possible to do with `deck-gl`?

**Comment by Pessimistress** [comment]
You can supply your own [renderSubLayers](https://deck.gl/docs/api-reference/geo-layers/tile-layer#rendersublayers) prop to the `MVTLayer` and render a `HeatmapLayer` instead.

However, `HeatmapLayer` is not really designed to work with tiled data, so you will likely see seams where the tiles join. Alternatively, you can render an independent `HeatmapLayer` and update its `data` prop when tiles are loaded. You can do so by subscribing to the [onViewportLoad](https://deck.gl/docs/api-reference/geo-layers/tile-layer#onviewportload) callback of the `MVTLayer`.

**Comment by makeupsomething** [comment]
Thank you for the suggestions!

I have been trying to use the heatmap layer with the data I get from my MVT tiles but it seems like the coordinate system is not working well with heatmap. I get coordinates from my MVT tile geometry like `[0.3515625, 0.750244140625]` but heatmap does not render anything for those coordinates.

**Comment by Pessimistress** [comment]
The MVT tiles are parsed into normalized coordinates. `renderSubLayers` receives props including `coordinateSystem`, `coordinateOrigin` and `modelMatrix` that also need to be passed to the HeatmapLayer.

**Reply by anne030303** [reply]
Hi! I have got the same problem. And I found different tile in renderSubLayers has its own coordinateSystem, coordinateOrigin and modelMatrix, how can I transform coordinates and merge them for HeatmapLayer? Thank you!

**Comment by makeupsomething** [comment]
Thank you, I wasn't aware of those props!

I still cannot get the heatmap to appear though. But I think it must be doing something because I can feel the measure dropping when zooming in and out.

Here is a code snippit
```
new MVTLayer({
    id: 'mvt-layer-2',
    data: `my/api/...`,
    loadOptions: {
        headers: {
            Authorization: 'Bearer ' + authToken.token,
        },
    },
    renderSubLayers: (props: any) => {
        const {
            data,
            coordinateOrigin,
            coordinateSystem,
            modelMatrix,
        } = props;

        return new HeatmapLayer({
            id: `${props.id}-heatmap-layer`,
            data,
            coordinateSystem: coordinateSystem,
            coordinateOrigin: coordinateOrigin,
            modelMatrix: modelMatrix,
            getPosition: (d) => d.geometry.coordinates,
            getWeight: (d) => {
                d.properties.Total
            },
        });
    },
})
```

Note that this code works when I replace `HeatmapLayer` with any other aggregation layer such as the `ScreenGridLayer`

**Reply by shaunakv1** [reply]
@makeupsomething  were you able to resolve the problem with the coordinate system? I am running into the same issue.

**Reply by makeupsomething** [reply]
No I did not manage to get this working. I ended up using a regular mapbox heatmap layer with `react-map-gl`.
